### PR TITLE
Extract nullable state operations to extension methods.

### PR DESCRIPTION
**Pull request by gafter**
Add tests to demonstrate that they are associative and commutative.
Add (skipped) tests to demonstrate that we have inconsistent meet and join functions.
    We will have to decide how to handle the differences.
I have not yet extracted the NullableState operations from ConversionsBase
    (e.g. testing if one NullableState is higher or lower than another in the lattice)
Also removed some inoperative code in MethodTypeInference
This relates to #31160

### CODE REVIEW

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 1]
```diff
@@ -4758,14 +4693,14 @@ protected override bool IntersectWith(ref LocalState self, ref LocalState other)
...
             }
-
-            bool isPossiblyNullableReferenceTypeTypeParameter(int slot)
-            {
-                Symbol symbol = variableBySlot[slot].Symbol;
-                return (object)symbol != null && VariableType(symbol).TypeSymbol?.IsPossiblyNullableReferenceTypeTypeParameter() == true;
-            }
         }
 
+        private static Func<(int slot, NullableWalker self), bool> isPossiblyNullableReferenceTypeTypeParameterDelegate = args =>
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >private [](start = 8, length = 7)

`readonly`

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs [hunk 2]
```diff
@@ -309,104 +525,21 @@ internal TypeSymbolWithAnnotations MergeNullability(TypeSymbolWithAnnotations ot
...
-                        Debug.Assert(a.IsAnyNullable() != b.IsAnyNullable());
-                    }
-
-                    hadNullabilityMismatch = true;
-                    return NullableAnnotation.NotNullable;
+                    throw ExceptionUtilities.UnexpectedValue(variance);
             }
         }
 
+        private static Func<TypeSymbol, bool> _IsPossiblyNullableReferenceTypeTypeParameterDelegate = type => type.IsPossiblyNullableReferenceTypeTypeParameter();
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs]: >private [](start = 8, length = 7)

`readonly`

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs [hunk 3]
```diff
@@ -31,6 +31,222 @@ public static bool IsAnyNotNullable(this NullableAnnotation annotation)
...
+            {
+                return selfAnnotation == NullableAnnotation.NotNullable || otherAnnotation == NullableAnnotation.NotNullable ?
+                            NullableAnnotation.NotNullable : NullableAnnotation.NotNullableBasedOnAnalysis;
+            }
+        }
+
+        /// <summary>
+        /// Meet two nullable annotations for computing the nullable annotation of a type parameter from upper bounds.
+        /// </summary>
+        public static NullableAnnotation MeetForFixingUpperBounds<T>(this NullableAnnotation a, NullableAnnotation b, T type, Func<T, bool> isPossiblyNullableReferenceTypeTypeParameter)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs]: >isPossiblyNullableReferenceTypeTypeParameter [](start = 140, length = 44)

Not used.

Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs]: >type [](start = 120, length = 4)

Not used.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 4]
```diff
@@ -57132,5 +57132,237 @@ static void F(object? x)
                 //         y = (object)x;
                 Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "(object)x").WithLocation(8, 13));
         }
+
+        [Fact]
+        public void TestJoinForFixingLowerBoundsIsAssociative()
+        {
+            var allAnnotations = (NullableAnnotation[])Enum.GetValues(typeof(NullableAnnotation));
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >(NullableAnnotation[])Enum.GetValues(typeof(NullableAnnotation)); [](start = 33, length = 65)

Consider extracting a helper method.

**Comment by gafter** [comment]
@cston Do you have any other comments?

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs [hunk 5]
```diff
@@ -31,6 +31,222 @@ public static bool IsAnyNotNullable(this NullableAnnotation annotation)
...
+            {
+                return a;
+            }
+
+            if (a.IsAnyNullable() && b.IsAnyNullable())
+            {
+                return NullableAnnotation.Nullable;
+            }
+
+            // If nullability on both sides matches - result is that nullability (trivial cases like these are handled before the switch)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs]: >the switch [](start = 126, length = 10)

Consider adjusting the comment, the ```switch``` is no longer part of the implementation.

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs [hunk 6]
```diff
@@ -31,6 +31,222 @@ public static bool IsAnyNotNullable(this NullableAnnotation annotation)
...
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs]: >the switch [](start = 126, length = 10)

Consider adjusting the comment, the switch is no longer part of the implementation.

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs [hunk 7]
```diff
@@ -31,6 +31,222 @@ public static bool IsAnyNotNullable(this NullableAnnotation annotation)
...
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs]: >the switc [](start = 126, length = 9)

Consider adjusting the comment, the switch is no longer part of the implementation.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 8]
```diff
@@ -4688,57 +4673,7 @@ protected override bool IntersectWith(ref LocalState self, ref LocalState other)
                 {
                     NullableAnnotation selfAnnotation = self[slot];
                     NullableAnnotation otherAnnotation = other[slot];
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >NullableAnnotation otherAnnotation = other[slot]; [](start = 20, length = 49)

Consider inlining the local.

**Pull request merged**